src/dataflow/core/type_inference.py

```diff
@@ -0,0 +1,351 @@
+from dataflow.core.program import (
+    BuildStructOp,
+    CallLikeOp,
+    Expression,
+    Program,
+    TypeName,
+    ValueOp,
+    roots_and_reentrancies,
+)
+
+
+@dataclass(frozen=True)
+class TypeInferenceError(Exception):
+    msg: str
+
+
+class Type(ABC):
+    pass
+
+
+class TypeVariable(Type):
+    pass
+
+
+class NamedTypeVariable(TypeVariable):
+    """A named type variable like T. Note that named type variables have scope:
+    several different functions might use the name T but they are distinct type
+    variables. As such, reference identity for NamedTypeVariables matters."""
+
+    name: str
+
+    def __init__(self, name: str):
+        self.name = name
+
+    def __repr__(self):
+        return f"$_{self.name}"
+
+
+class AnonTypeVariable(TypeVariable):
+    """An anonymous type variable used to represent types that need to be inferred.
+    Compared by reference equality."""
+
+    pass
+
+    def __repr__(self):
+        return f"$_{hex(id(self))}"
+
+
+@dataclass(frozen=True)
+class TypeApplication(Type):
+    """A type constructor with some arguments."""
+
+    constructor: str
+    args: Sequence[Type] = field(default_factory=list)
+
+    def __repr__(self):
+        if len(self.args) == 0:
+            return f"{self.constructor}"
+        else:
+            return (
+                f'{self.constructor}[{",".join([arg.__repr__() for arg in self.args])}]'
+            )
+
+
+@dataclass(frozen=False)
+class Computation:
+    """A representation of a single function application during type inference.
+    Value literals are treated as functions from empty arguments to their value."""
+
+    # The name of the function or value literal.
+    op: Optional[str]
+    # For mapping back to Expression.id
+    id: str
+    # The arguments to the function
+    args: List["Computation"]
+    # Declared type arguments based on the signature of op (defined in a Definition).
+    type_args: List[NamedTypeVariable]
+    # The type of this function, always represented as a
+    # TypeApplication("Lambda", arg_types + [return_type])
+    # Mutated during type inference.
+    type: TypeApplication
+
+    @property
+    def return_type(self) -> Type:
+        """The output type of this function. Mutated during type inference"""
+        return self.type.args[-1]
+
+
+def infer_types(program: Program, library: Dict[str, Definition]) -> Program:
+    """Main entry point of a Hindley-Milner like inference algorithm.
+    The high-level algorithm is :
+    • Convert each Expression in `program` to a Computation, which in turn represents
+      each function invocation as a TypeApplication of the form
+      Lambda[AnonVariable1, AnonVariable12..., Output],
+      where Output is Expression.type if defined or an AnonymousTypeVariable otherwise.
+    • Unify (see _unify) the Definition of the function (from `library`) with the
+      Computation. Each unification mutable updates the substitutions map,
```

Comment:
I think this doc string would be clearer if you mention earlier that a global substitutions map is mutated throughout the inference and maintains the current bounds on the type variables.